### PR TITLE
Add switching to previous workspace

### PR DESCRIPTION
**Pull request by CharlesEkkel**
#### Describe your PR, what does it fix/add?
Add the ability to switch to the last visited workspace as described in #324, inspired by i3. The syntax suggested in the issue has been adopted: `bind=MODIFIER,KEY,workspace,previous`.

Also included is a toggle setting `general:auto_back_and_forth` to allow switching back to the previous workspace by trying to switch to the same workspace again. I use it all the time in sway and my workflow would be _greatly_ improved by the ability to quickly switch to and from a workspace to check on email, discord, spotify, etc. 
E.g. start in workspace 1, then `Super+7` to workspace 7, and then `Super+7` again will move to workspace 1. 

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)
I've moved a lot of code from `CKeybindManager::changeworkspace` to `CCompositor::changeWorkspace`, to make it easier to reuse the raw workspace switching.

This is also my first pull request, so my apologies if I've broken any conventions, just let me know and I'll fix it. I'm also not super experienced with C++, but I think I worked out enough for this small change.

#### Is it ready for merging, or does it need work?
Yes it's ready as far as I know (it builds and works in debug mode). However:

- I'm not sure if I've put `m_bAutoBackAndForth` in the right place - let me know if there's somewhere else for settings like that to go.
- Not sure if some hyprctl commands need to be added too.
- Of course wiki updates will be necessary too, but I don't know how that's normally done.

### CODE REVIEW

File: src/Compositor.cpp [hunk 1]
```diff
@@ -280,6 +280,9 @@ void CCompositor::startCompositor() {
 
     wlr_xcursor_manager_set_cursor_image(m_sWLRXCursorMgr, "left_ptr", m_sWLRCursor);
 
+    // TODO: Is this the ideal place to put this?
+    m_bAutoBackAndForth = g_pConfigManager->getConfigValuePtr("general:auto_back_and_forth")->intValue;
```
Comment (vaxerski) [inline src/Compositor.cpp:284]: wtf

File: src/Compositor.cpp [hunk 2]
```diff
@@ -722,6 +725,163 @@ CWorkspace* CCompositor::getWorkspaceByID(const int& id) {
...
 
+/**
+ * @brief Switch to a target workspace, creating a new one on the current 
+ * monitor if required.
+ * 
+ * @param TENTATIVE_TARGET_ID ID of the target workspace.
+ * @param TARGET_WORKSPACE_NAME Name of target workspace, only needed if the 
+ * workspace doesn't yet exist.
+ */
+void CCompositor::changeWorkspace(const int TENTATIVE_TARGET_ID, const std::string& TARGET_WORKSPACE_NAME) {
```
Comment (vaxerski) [inline src/Compositor.cpp:736]: why did you move it?

File: src/Compositor.hpp [hunk 3]
```diff
@@ -85,11 +85,15 @@ class CCompositor {
     wlr_surface*            m_pLastFocus = nullptr;
     CWindow*                m_pLastWindow = nullptr;
     SMonitor*               m_pLastMonitor = nullptr;
+    int                     m_iLastWorkspaceID = -1; // Start with no last workspace, represented by -1.
```
Comment (vaxerski) [inline src/Compositor.hpp:88]: why?? is it not better to make a chain??

**Comment by vaxerski** [comment]
generally 90% of the code here is just a big ????

Why are you storing the last workspace in CCompositor?

Why are you moving the function?

Why did you *refactor* it???

this entire thing can be done with a chain in 10 lines of code or less.

**Comment by CharlesEkkel** [comment]
> generally 90% of the code here is just a big ????
> 
> Why are you storing the last workspace in CCompositor?

I stored it there because I figured it was information scoped to the compositor, not workspaces or even monitors (previous workspace may be on another monitor). But you have a point, it might make more sense to implement as a chain for a series of 'back' operations. In which case, I could just store the source workspace in each workspace, instead? And use that to go back?

> Why are you moving the function?
> 
> Why did you _refactor_ it???

I wanted to re-use the workspace switching code and figured that it would make sense to go there, rather than in CKeybindManager. But I probably did unnecessarily change it too far, my bad, I didn't consider that would be an issue. I'll fix that, and just have conditions for setting the target workspace in CKeybindManager::changeworkspace instead. Much smaller change then.

> this entire thing can be done with a chain in 10 lines of code or less.

Probably. 

Look, sorry if you felt like I wasted your time. I'm excited to contribute here and I really like what you've made here. But I'm learning, and I really do want this feature. I'll try and fix it up as you suggest, just need some clarification:

1. Is it acceptable for me to store the previous workspace inside each workspace object?
2. Where can I store/access the `autoBackAndForth` setting?

File: src/Compositor.cpp (hunk 1 above)
Comment (CharlesEkkel) [reply src/Compositor.cpp:284]: Where can I read this in then? I haven't been able to determine how configuration parameters are read in and initialised.

File: src/Compositor.hpp (hunk 3 above)
Comment (CharlesEkkel) [reply src/Compositor.hpp:88]: I suppose so, I guess it's just never been relevant to my workflow. I think I must have assumed the `auto_back_and_forth` behaviour as the default.

In my comment below I suggest having each workspace remember the workspace it was accessed from, to form such a chain. Does that sound workable?

File: src/Compositor.hpp [hunk 4]
```diff
@@ -85,11 +85,15 @@ class CCompositor {
...
     
     SSeat                   m_sSeat;
 
     bool                    m_bReadyToProcess = false;
 
+    // Settings
+    bool                    m_bAutoBackAndForth = false;
```
[the remark this answers is not in the document]
Comment (CharlesEkkel) [reply src/Compositor.hpp:95]: Fair enough. Now that I think about it, should I just read the value from the config manager every time instead?

**Comment by vaxerski** [comment]
> doesn't currently exist, it should just be created rather than doing nothing

no

> This implementation will also allow for cycles to occur, unless I destroy (set to -1) the 'previous workspace' id each time the user goes back a workspace.

make it configurable

**Comment by yavko** [comment]
Any progress on this?

**Comment by CharlesEkkel** [comment]
> Any progress on this?

Shoot, thanks for the reminder! I completely forgot about this. I'm starting a new job tomorrow so I'll try and redo this PR over the next few days or at the latest next weekend. If anyone else wants to give it a go instead - please, feel free. Can't be hard to do better than I did 😅

**Comment by yavko** [comment]
> > Any progress on this?
> 
> Shoot, thanks for the reminder! I completely forgot about this. I'm starting a new job tomorrow so I'll try and redo this PR over the next few days or at the latest next weekend. If anyone else wants to give it a go instead - please, feel free. Can't be hard to do better than I did sweat_smile

Good luck on the job!

**Comment by CharlesEkkel** [comment]
@yavko Thanks! It's been a good first week, lots of flexibility to work from home has been lovely.

@vaxerski I've tried to take on your feedback as best I could. The only thing I didn't do was ignore non-existent workspaces when going to previous. I checked in i3 & sway, and this is how they do it too - users who want this setting will expect the behaviour I've implemented (automatically re-creating previous workspaces which have been destroyed).

The below PR description has been updated.

### Describe your PR, what does it fix/add?

Add the ability to switch to the last visited workspace as described in https://github.com/hyprwm/Hyprland/issues/324, inspired by i3. The syntax suggested in the issue has been adopted: `bind=MODIFIER,KEY,workspace,previous`

Also included is a toggle setting `general:workspace_back_and_forth` to allow switching back to the previous workspace by trying to switch to the same workspace again. I use it all the time in sway and my workflow would be greatly improved by the ability to quickly switch to and from a workspace to check on email, discord, spotify, etc.

E.g. start in workspace 1, then Super+7 to workspace 7, and then Super+7 again will move to workspace 1.

Finally there's a `general:allow_workspace_cycles` option which toggles whether or not the previous workspace ID is retained after switching to the previous workspace. If this is not done, then there are situations in which a cycle can be formed and the previous workspace can be switched to endlessly. 

### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)

Nope, besides some potentially unsafe pointer use, but as far as I can tell it's inline with the rest of the surrounding code and it would appear to be safe as far as I can tell. No bugs found in my testing, it's not a huge change.

My only real caveat is that today I've been getting an _"Attempted to draw NULL texture!"_ assertion error when I run in debug, which gives me a black screen when I ignore it - nonetheless I've still been able to check that my code works by just looking at the live logs. So it should be fine as I suspect the issue is that there's something up with my dev environment.

### Is it ready for merging, or does it need work?

Yes it's ready, pending approval.

Of course wiki updates will be necessary too, but I don't know how that's normally done.

File: src/helpers/Workspace.hpp [hunk 5]
```diff
@@ -18,6 +18,9 @@ class CWorkspace {
     int             m_iID = -1;
     std::string     m_szName = "";
     uint64_t        m_iMonitorID = -1;
+    // Previous workspace ID is stored during a workspace change, allowing travel
+    // to the previous workspace.
+    int m_iPrevWorkspaceID = -1;
```
Comment (vaxerski) [inline src/helpers/Workspace.hpp:23]: code styling: indent

File: src/managers/KeybindManager.cpp [hunk 6]
```diff
@@ -454,11 +454,33 @@ void CKeybindManager::changeworkspace(std::string args) {
...
+            Debug::log(LOG, "No previous workspace to change to");
+            return;
+        }
+        else {
+            workspaceToChangeTo = P_CURRENT_WORKSPACE->m_iPrevWorkspaceID;
+            isSwitchingToPrevious = true;
+
+            // If the previous workspace ID isn't reset, cycles can form when continually going
+            // to the previous workspace again and again.
+            if (!g_pConfigManager->getConfigValuePtr("general:allow_workspace_cycles")->intValue)
```
Comment (vaxerski) [inline src/managers/KeybindManager.cpp:481]: no, defeats the purpose of getConfigValuePtr.

```
static auto *const PALLOWWORKSPACECYCLES = &g_pConfigManager->getConfigValuePtr("general:allow_workspace_cycles")->intValue;
```
on top of the func, then `*PALLOWWORKSPACECYCLES`

File: src/managers/KeybindManager.cpp [hunk 7]
```diff
@@ -468,6 +490,23 @@ void CKeybindManager::changeworkspace(std::string args) {
         return;
     }
 
+    // Workspace_back_and_forth being enabled means that an attempt to switch to 
+    // the current workspace will instead switch to the previous.
+    if (g_pConfigManager->getConfigValuePtr("general:workspace_back_and_forth")->intValue == 1
```
Comment (vaxerski) [inline src/managers/KeybindManager.cpp:495]: same as above

File: src/managers/KeybindManager.cpp [hunk 8]
```diff
@@ -468,6 +490,23 @@ void CKeybindManager::changeworkspace(std::string args) {
...
+        
+        const auto P_CURRENT_WORKSPACE = g_pCompositor->getWorkspaceByID(
+            g_pCompositor->getMonitorFromCursor()->activeWorkspace);
+
+        workspaceToChangeTo = P_CURRENT_WORKSPACE->m_iPrevWorkspaceID;
+        isSwitchingToPrevious = true;
+
+        // If the previous workspace ID isn't reset, cycles can form when continually going
+        // to the previous workspace again and again.
+        if (!g_pConfigManager->getConfigValuePtr("general:allow_workspace_cycles")->intValue)
```
Comment (vaxerski) [inline src/managers/KeybindManager.cpp:506]: same as above

File: src/managers/KeybindManager.cpp [hunk 9]
```diff
@@ -477,6 +516,10 @@ void CKeybindManager::changeworkspace(std::string args) {
 
         const auto PWORKSPACETOCHANGETO = g_pCompositor->getWorkspaceByID(workspaceToChangeTo);
 
+        if (!isSwitchingToPrevious)
+            // Remember previous workspace.
+            PWORKSPACETOCHANGETO->m_iPrevWorkspaceID = g_pCompositor->getMonitorFromCursor()->activeWorkspace;
```
Comment (vaxerski) [inline src/managers/KeybindManager.cpp:521]: `g_pCompositor->m_pLastMonitor`

File: src/managers/KeybindManager.cpp [hunk 10]
```diff
@@ -454,11 +454,33 @@ void CKeybindManager::changeworkspace(std::string args) {
...
+    // to a previous workspace.
+    bool isSwitchingToPrevious = false;
+
     if (args.find("[internal]") == 0) {
         workspaceToChangeTo = std::stoi(args.substr(10));
         const auto PWORKSPACE = g_pCompositor->getWorkspaceByID(workspaceToChangeTo);
         if (PWORKSPACE)
             workspaceName = PWORKSPACE->m_szName;
+    } else if (args.find("previous") == 0) {
+        const auto P_CURRENT_WORKSPACE = g_pCompositor->getWorkspaceByID(
```
Comment (vaxerski) [inline src/managers/KeybindManager.cpp:467]: code style, no snake case

File: src/config/ConfigManager.cpp [hunk 11]
```diff
@@ -34,6 +34,8 @@ void CConfigManager::setDefaultVars() {
     configValues["general:apply_sens_to_raw"].intValue = 0;
     configValues["general:main_mod"].strValue = "SUPER";                                               // exposed to the user for easier configuring
     configValues["general:main_mod_internal"].intValue = g_pKeybindManager->stringToModMask("SUPER");  // actually used and automatically calculated
+    configValues["general:workspace_back_and_forth"].intValue = 0;
```
Comment (vaxerski) [inline src/config/ConfigManager.cpp:37]: if

> Also included is a toggle setting general:workspace_back_and_forth to allow switching back to the previous workspace by trying to switch to the same workspace again. I use it all the time in sway and my workflow would be greatly improved by the ability to quickly switch to and from a workspace to check on email, discord, spotify, etc.

wouldn't it be appropriate to put this in `binds:`?

File: src/helpers/Workspace.hpp (hunk 5 above)
Comment (CharlesEkkel) [reply src/helpers/Workspace.hpp:23]: Thanks.

File: src/managers/KeybindManager.cpp (hunk 6 above)
Comment (CharlesEkkel) [reply src/managers/KeybindManager.cpp:481]: Will do.

File: src/managers/KeybindManager.cpp (hunk 10 above)
Comment (CharlesEkkel) [reply src/managers/KeybindManager.cpp:467]: Will do, although I'd point out that the all-caps variables can get a bit hard to read.

File: src/config/ConfigManager.cpp (hunk 11 above)
Comment (CharlesEkkel) [reply src/config/ConfigManager.cpp:37]: Good idea.

**Comment by CharlesEkkel** [comment]
@vaxerski Adjustments made.

File: src/managers/KeybindManager.cpp [hunk 12]
```diff
@@ -454,11 +454,34 @@ void CKeybindManager::changeworkspace(std::string args) {
...
+    bool isSwitchingToPrevious = false;
+
     if (args.find("[internal]") == 0) {
         workspaceToChangeTo = std::stoi(args.substr(10));
         const auto PWORKSPACE = g_pCompositor->getWorkspaceByID(workspaceToChangeTo);
         if (PWORKSPACE)
             workspaceName = PWORKSPACE->m_szName;
+    } else if (args.find("previous") == 0) {
+        const auto PCURRENTWORKSPACE = g_pCompositor->getWorkspaceByID(
+            g_pCompositor->getMonitorFromCursor()->activeWorkspace);
```
Comment (vaxerski) [inline src/managers/KeybindManager.cpp:468]: `g_pCompositor->m_pLastMonitor`

Comment (CharlesEkkel) [reply src/managers/KeybindManager.cpp:468]: Oh crap sorry.

**Comment by vaxerski** [comment]
For the wiki, since you probably know best how this works, make a PR to [the wiki](https://github.com/hyprwm/hyprland-wiki)

**Comment by KulkarniKaustubh** [comment]
Hey! Is there any way to bind `workspace_back_and_forth` to something?

For example:
`bind = $mainMod, TAB, workspace_back_and_forth,`

**Comment by csdvrx** [comment]
> Hey! Is there any way to bind workspace_back_and_forth to something?

You need to setup this general option, it's not a dispatcher.

Use the instead:

`bind = $mainMod, TAB,workspace, previous`

**Comment by KulkarniKaustubh** [comment]
> > Hey! Is there any way to bind workspace_back_and_forth to something?
> 
> You need to setup this general option, it's not a dispatcher.
> 
> Use this instead:
> 
> `bind = $mainMod, TAB,workspace, previous`

Thank you! This was exactly what I wanted :)
